File: Backtracking/leetcode_40.cpp
// Given a collection of candidate numbers (candidates) and a target number (target), find all unique 
// combinations in candidates where the candidate numbers sum to target.

// Each number in candidates may only be used once in the combination.

// Note: The solution set must not contain duplicate combinations.

#include<bits/stdc++.h>
using namespace std;


class Solution {
public:
    vector<vector<int>> result;
    
    void comsum(vector<int> &curr, int target, int sum, vector<int> &candidates, int curInd, int n){
        if(target == sum){
            result.push_back(curr);
            return;
        }
        else if(sum>target){
            return;
        }
        
        for(int i = curInd; i < n; i++){
            if(i != curInd && candidates[i]==candidates[i-1])               //to avoid picking up the same combnations i.e. we don't pick same element for certain kth position of a combination 
                continue;
            sum += candidates[i];
            curr.push_back(candidates[i]);
            comsum(curr, target, sum, candidates, i+1, n);
            sum -= candidates[i];
            curr.pop_back();
        }
        
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> curr;
        int n = candidates.size();
        sort(candidates.begin(), candidates.end());
        comsum(curr, target, 0, candidates, 0, n);
        return result;
    }
};

int main()
{
    vector<int> candidates = {10, 1, 2, 7, 6, 1, 5}; // Example candidate set
    int target = 8;                                  // Example target sum

    Solution solution;
    vector<vector<int>> combinations = solution.combinationSum2(candidates, target);

    cout << "All unique combinations that sum up to " << target << ":" << endl;
    for (const vector<int> &combination : combinations)
    {
        for (int num : combination)
        {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Backtracking/leetcode_40.py
# Problem Link ====>>https://leetcode.com/problems/combination-sum-ii/description/

class Solution:
    def __init__(self):
        self.result = []  # List to store unique combinations

    def comsum(self, curr, target, sum, candidates, curInd, n):
        if target == sum:
            self.result.append(curr.copy())  # Append a copy of the combination
            return

        if sum > target:
            return

        for i in range(curInd, n):
            # Avoid duplicates
            if i > curInd and candidates[i] == candidates[i - 1]:
                continue

            sum += candidates[i]
            curr.append(candidates[i])
            self.comsum(curr, target, sum, candidates, i + 1, n)
            sum -= candidates[i]
            curr.pop()

    def combinationSum2(self, candidates, target):
        curr = []  # Temporary list for combinations
        n = len(candidates)
        candidates.sort()  # Sort for duplicate handling
        self.comsum(curr, target, 0, candidates, 0, n)
        return self.result


if __name__ == "__main__":
    candidates = [10, 1, 2, 7, 6, 1, 5]  # Example candidate set
    target = 8  # Example target sum

    solution = Solution()
    combinations = solution.combinationSum2(candidates, target)

    print("All unique combinations that sum up to", target, ":")
    for combination in combinations:
        print(combination)
